File: Src/java/engine-fhir/src/test/java/org/opencds/cqf/cql/engine/fhir/data/FhirExecutionTestBase.java
package org.opencds.cqf.cql.engine.fhir.data;

import ca.uhn.fhir.context.FhirContext;
import ca.uhn.fhir.context.FhirVersionEnum;
import jakarta.xml.bind.JAXBException;
import java.io.File;
import java.io.IOException;
import java.net.URLDecoder;
import java.util.ArrayList;
import org.cqframework.cql.cql2elm.*;
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider;
import org.cqframework.cql.cql2elm.tracking.TrackBack;
import org.fhir.ucum.UcumException;
import org.hl7.elm.r1.Library;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.opencds.cqf.cql.engine.data.CompositeDataProvider;
import org.opencds.cqf.cql.engine.execution.CqlEngine;
import org.opencds.cqf.cql.engine.execution.Environment;
import org.opencds.cqf.cql.engine.fhir.model.*;
import org.opencds.cqf.cql.engine.fhir.retrieve.RestFhirRetrieveProvider;
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver;

public abstract class FhirExecutionTestBase {
    public LibraryManager getLibraryManager() {
        return libraryManager;
    }

    public ModelManager getModelManager() {
        return modelManager;
    }

    public Environment getEnvironment() {
        return new Environment(getLibraryManager());
    }

    public CqlEngine getEngine() {
        return new CqlEngine(getEnvironment());
    }

    private static LibraryManager libraryManager;
    private static ModelManager modelManager;

    protected static Dstu2FhirModelResolver dstu2ModelResolver;
    protected static RestFhirRetrieveProvider dstu2RetrieveProvider;
    protected static CompositeDataProvider dstu2Provider;

    protected static Dstu3FhirModelResolver dstu3ModelResolver;
    protected static RestFhirRetrieveProvider dstu3RetrieveProvider;
    protected static CompositeDataProvider dstu3Provider;

    protected static R4FhirModelResolver r4ModelResolver;
    protected static RestFhirRetrieveProvider r4RetrieveProvider;
    protected static CompositeDataProvider r4Provider;

    Library library = null;
    // protected File xmlFile = null;

    @BeforeAll
    public static void setup() {
        FhirContext dstu2Context = FhirContext.forCached(FhirVersionEnum.DSTU2);
        dstu2ModelResolver = new CachedDstu2FhirModelResolver();
        dstu2RetrieveProvider = new RestFhirRetrieveProvider(
                new SearchParameterResolver(dstu2Context),
                dstu2ModelResolver,
                dstu2Context.newRestfulGenericClient("http://fhirtest.uhn.ca/baseDstu2"));
        dstu2Provider = new CompositeDataProvider(dstu2ModelResolver, dstu2RetrieveProvider);

        FhirContext dstu3Context = FhirContext.forCached(FhirVersionEnum.DSTU3);
        dstu3ModelResolver = new CachedDstu3FhirModelResolver();
        dstu3RetrieveProvider = new RestFhirRetrieveProvider(
                new SearchParameterResolver(dstu3Context),
                dstu3ModelResolver,
                dstu3Context.newRestfulGenericClient("http://measure.eval.kanvix.com/cqf-ruler/baseDstu3"));
        dstu3Provider = new CompositeDataProvider(dstu3ModelResolver, dstu3RetrieveProvider);

        FhirContext r4Context = FhirContext.forCached(FhirVersionEnum.R4);
        r4ModelResolver = new CachedR4FhirModelResolver();
        r4RetrieveProvider = new RestFhirRetrieveProvider(
                new SearchParameterResolver(r4Context),
                r4ModelResolver,
                r4Context.newRestfulGenericClient("http://measure.eval.kanvix.com/cqf-ruler/baseDstu4"));
        r4Provider = new CompositeDataProvider(r4ModelResolver, r4RetrieveProvider);

        modelManager = new ModelManager();
        var compilerOptions = CqlCompilerOptions.defaultOptions();
        libraryManager = new LibraryManager(modelManager, compilerOptions);
        libraryManager.getLibrarySourceLoader().clearProviders();
        libraryManager.getLibrarySourceLoader().registerProvider(new FhirLibrarySourceProvider());
        libraryManager.getLibrarySourceLoader().registerProvider(new TestLibrarySourceProvider());
    }

    @BeforeEach
    public void beforeEachTestMethod() throws JAXBException, IOException, UcumException {
        String fileName = this.getClass().getSimpleName();
        if (library == null) {
            try {
                File cqlFile = new File(URLDecoder.decode(
                        this.getClass().getResource(fileName + ".cql").getFile(), "UTF-8"));

                CqlCompiler compiler = new CqlCompiler(getLibraryManager());

                var library = compiler.run(cqlFile);

                if (!compiler.getErrors().isEmpty()) {
                    System.err.println("Translation failed due to errors:");
                    ArrayList<String> errors = new ArrayList<>();
                    for (CqlCompilerException error : compiler.getErrors()) {
                        TrackBack tb = error.getLocator();
                        String lines = tb == null
                                ? "[n/a]"
                                : String.format(
                                        "[%d:%d, %d:%d]",
                                        tb.getStartLine(), tb.getStartChar(), tb.getEndLine(), tb.getEndChar());
                        System.err.printf("%s %s%n", lines, error.getMessage());
                        errors.add(lines + error.getMessage());
                    }
                    throw new IllegalArgumentException(errors.toString());
                }

                this.library = library;

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static org.hl7.elm.r1.VersionedIdentifier toElmIdentifier(String name) {
        return new org.hl7.elm.r1.VersionedIdentifier().withId(name);
    }

    public static org.hl7.elm.r1.VersionedIdentifier toElmIdentifier(String name, String version) {
        return new org.hl7.elm.r1.VersionedIdentifier().withId(name).withVersion(version);
    }
}

File: Src/java/engine-fhir/src/test/kotlin/org/opencds/cqf/cql/engine/fhir/data/FhirExecutionTestBase.kt
package org.opencds.cqf.cql.engine.fhir.data

import ca.uhn.fhir.context.FhirContext
import ca.uhn.fhir.context.FhirVersionEnum
import java.io.IOException
import kotlinx.io.files.Path
import org.cqframework.cql.cql2elm.CqlCompiler
import org.cqframework.cql.cql2elm.CqlCompilerException.ErrorSeverity
import org.cqframework.cql.cql2elm.CqlCompilerOptions.Companion.defaultOptions
import org.cqframework.cql.cql2elm.LibraryManager
import org.cqframework.cql.cql2elm.ModelManager
import org.cqframework.cql.cql2elm.quick.FhirLibrarySourceProvider
import org.fhir.ucum.UcumException
import org.hl7.elm.r1.Library
import org.hl7.elm.r1.VersionedIdentifier
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.BeforeEach
import org.opencds.cqf.cql.engine.data.CompositeDataProvider
import org.opencds.cqf.cql.engine.execution.CqlEngine
import org.opencds.cqf.cql.engine.execution.Environment
import org.opencds.cqf.cql.engine.fhir.model.CachedDstu2FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.model.CachedDstu3FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.model.CachedR4FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.model.Dstu2FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.model.Dstu3FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.model.R4FhirModelResolver
import org.opencds.cqf.cql.engine.fhir.retrieve.RestFhirRetrieveProvider
import org.opencds.cqf.cql.engine.fhir.searchparam.SearchParameterResolver

abstract class FhirExecutionTestBase {
    val libraryManager: LibraryManager
        get() = Companion.libraryManager!!

    val modelManager: ModelManager
        get() = Companion.modelManager!!

    val environment: Environment
        get() = Environment(this.libraryManager)

    private var _engine: CqlEngine? = null
    val engine: CqlEngine
        get() {
            if (_engine == null) {
                _engine = CqlEngine(this.environment)
            }
            return _engine!!
        }

    var library: Library? = null

    @BeforeEach
    @Throws(IOException::class, UcumException::class)
    fun beforeEachTestMethod() {
        val fileName = this.javaClass.getSimpleName()
        if (library == null) {
            try {
                val cqlFile = Path(this.javaClass.getResource("$fileName.cql")!!.path)

                val compiler = CqlCompiler(libraryManager = libraryManager)

                val library = compiler.run(cqlFile)

                val errors = compiler.exceptions.filter { it.severity == ErrorSeverity.Error }
                if (errors.isNotEmpty()) {
                    System.err.println("Translation failed due to errors:")
                    val messages = mutableListOf<String>()
                    for (error in errors) {
                        val tb = error.locator
                        val lines =
                            if (tb == null) "[n/a]"
                            else "[${tb.startLine}:${tb.startChar}, ${tb.endLine}:${tb.endChar}]"
                        System.err.printf("%s %s%n", lines, error.message)
                        messages.add(lines + error.message)
                    }
                    throw IllegalArgumentException(messages.toString())
                }

                this.library = library
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    companion object {
        private var libraryManager: LibraryManager? = null
        private var modelManager: ModelManager? = null

        @JvmStatic protected var dstu2ModelResolver: Dstu2FhirModelResolver? = null
        protected var dstu2RetrieveProvider: RestFhirRetrieveProvider? = null
        @JvmStatic protected var dstu2Provider: CompositeDataProvider? = null

        @JvmStatic protected var dstu3ModelResolver: Dstu3FhirModelResolver? = null
        @JvmStatic protected var dstu3RetrieveProvider: RestFhirRetrieveProvider? = null
        @JvmStatic protected var dstu3Provider: CompositeDataProvider? = null

        @JvmStatic protected var r4ModelResolver: R4FhirModelResolver? = null
        protected var r4RetrieveProvider: RestFhirRetrieveProvider? = null
        @JvmStatic protected var r4Provider: CompositeDataProvider? = null

        @JvmStatic
        // protected File xmlFile = null;
        @BeforeAll
        fun setup() {
            val dstu2Context = FhirContext.forCached(FhirVersionEnum.DSTU2)
            dstu2ModelResolver = CachedDstu2FhirModelResolver()
            dstu2RetrieveProvider =
                RestFhirRetrieveProvider(
                    SearchParameterResolver(dstu2Context),
                    dstu2ModelResolver!!,
                    dstu2Context.newRestfulGenericClient("http://fhirtest.uhn.ca/baseDstu2"),
                )
            dstu2Provider = CompositeDataProvider(dstu2ModelResolver, dstu2RetrieveProvider)

            val dstu3Context = FhirContext.forCached(FhirVersionEnum.DSTU3)
            dstu3ModelResolver = CachedDstu3FhirModelResolver()
            dstu3RetrieveProvider =
                RestFhirRetrieveProvider(
                    SearchParameterResolver(dstu3Context),
                    dstu3ModelResolver!!,
                    dstu3Context.newRestfulGenericClient(
                        "http://measure.eval.kanvix.com/cqf-ruler/baseDstu3"
                    ),
                )
            dstu3Provider = CompositeDataProvider(dstu3ModelResolver, dstu3RetrieveProvider)

            val r4Context = FhirContext.forCached(FhirVersionEnum.R4)
            r4ModelResolver = CachedR4FhirModelResolver()
            r4RetrieveProvider =
                RestFhirRetrieveProvider(
                    SearchParameterResolver(r4Context),
                    r4ModelResolver!!,
                    r4Context.newRestfulGenericClient(
                        "http://measure.eval.kanvix.com/cqf-ruler/baseDstu4"
                    ),
                )
            r4Provider = CompositeDataProvider(r4ModelResolver, r4RetrieveProvider)

            modelManager = ModelManager()
            val compilerOptions = defaultOptions()
            libraryManager = LibraryManager(modelManager!!, compilerOptions)
            libraryManager!!.librarySourceLoader.clearProviders()
            libraryManager!!.librarySourceLoader.registerProvider(FhirLibrarySourceProvider())
            libraryManager!!.librarySourceLoader.registerProvider(TestLibrarySourceProvider())
        }

        fun toElmIdentifier(name: String?): VersionedIdentifier {
            return VersionedIdentifier().withId(name)
        }

        fun toElmIdentifier(name: String?, version: String?): VersionedIdentifier {
            return VersionedIdentifier().withId(name).withVersion(version)
        }
    }
}
